### PR TITLE
Wiegand driver for RIOT-OS

### PR DESCRIPTION
**Pull request by mxgxw**
### Wiegand driver for RIOT-OS

This commit adds support for the Wiegand protocol.
Wiegand is widely used protocol in RFID card readers, keyboards and several access control devices.
The driver adds a new folder under /drivers and updates all needed dependencies.
An example is given to test the implementation under /examples.

The example generates a wiegand frame and decodes it using the driver.
You'll need to use four GPIOs to test the implementation.

It was tested under the following boards:

- Arduino MKR1000
- Arduino Zero
- [Custom ATSAMR21 based board](https://www.hackerspace.sv/wiki/HSSV_ATSAMR21_Breakout_English)*

*This board is not commercially available but any ATSAMR21-based board
could be used for testing.

### Testing procedure

This example is valid for the Arduino-MKR1000

1. Place a jumper between A3 and A4 (Wiegand Data 0)
2. Place a jumper between A5 and A6 (Wiegand Data 1)
3. Run the command:
    BOARD=arduino-mkr1000 PORT=/dev/ttyACM0 make flash
4. Run the terminal to test the output:
    BOARD=arduino-mkr1000 PORT=/dev/ttyUSB0 make term

The example runs two RIOT threads. One of the threads generates a wiegand frame and the other decodes the data using the driver.

If the test is successful you should get the following output:
```
2019-09-12 18:58:51,826 - INFO # main(): This is RIOT! (Version: compilation)
2019-09-12 18:58:51,827 - INFO # Welcome to RIOT!
2019-09-12 18:58:51,828 - INFO # Starting Wiegand reader thread...
2019-09-12 18:58:51,830 - INFO # Starting Wiegand simulation thread...
2019-09-12 18:58:51,831 - INFO # Sending WG frame...
2019-09-12 18:58:51,832 - INFO # All up, running the shell now
> 2019-09-12 18:58:51,841 - INFO #  WG frame sent...
2019-09-12 18:58:56,846 - INFO # Sending WG frame...
2019-09-12 18:58:56,878 - INFO # WG frame sent...
2019-09-12 18:58:56,925 - INFO # Card found..
2019-09-12 18:58:56,926 - INFO # Card high: 0
2019-09-12 18:58:56,927 - INFO # Card low:  1943

```
It will repeat every 5 seconds.

### CODE REVIEW

File: drivers/include/wiegand.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * @brief   Device descriptor for a wiegand device
+   */
+  typedef struct {
+    uint32_t code;                         /**< Wiegand code */
+    int16_t type;                          /**< Wiegand type */
+    volatile uint32_t wg_card_temp_high;   /**< Card MSB */
+    volatile uint32_t wg_card_temp;        /**< Card LSV  */
+    volatile uint32_t wg_last_wiegand;     /**< Last card lecture */
+    volatile int16_t wg_bit_count;         /**< Number of bits read */
```
Comment (benpicco) [inline drivers/include/wiegand.h]: Why are these `volatile`?

File: drivers/include/wiegand.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * @brief   Initialize a Wiegand device
+   *
+   * @param[out] dev          device descriptor
+   * @param[in] params        configuration parameters
+   *
+   * @return                   0 on success
+   * @return                  -1 on error
+   */
+  int16_t wg_init(wiegand_t *dev, const wiegand_params_t *params);
```
Comment (benpicco) [inline drivers/include/wiegand.h]: I'd suggest you to use `wiegand_` as a prefix instead of `wg_`.
Also just use `int` for error codes.

File: drivers/wiegand/wiegand.c [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+ */
+
+#include "xtimer.h"
+#include "wiegand_params.h"
+
+/**
+ * @brief Handler for D0 signal in Wiegand protocol
+ *
+ */
+void wg_read_d0(void *arg)
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c]: private functions should be `static`.

File: drivers/wiegand/wiegand.c [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+  }
+  else
+  {
+    // D1 represent binary 1, so OR card data with 1 then
+    dev->wg_card_temp |= 1;
+    // left shift card data
+    dev->wg_card_temp <<= 1;
+  }
+  // Keep track of last wiegand bit received
+  dev->wg_last_wiegand = xtimer_now_usec64();
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c]: since `dev->wg_last_wiegand` is 32 bit anyway, `xtimer_now_usec()` is enough.

File: drivers/include/wiegand.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  } wiegand_params_t;
+
+  /**
+   * @brief   Device descriptor for a wiegand device
+   */
+  typedef struct {
+    uint32_t code;                         /**< Wiegand code */
+    int16_t type;                          /**< Wiegand type */
+    volatile uint32_t wg_card_temp_high;   /**< Card MSB */
+    volatile uint32_t wg_card_temp;        /**< Card LSV  */
```
Comment (benpicco) [inline drivers/include/wiegand.h]: Have you considered just using `uint64_t` here?

File: examples/wiegand/main.c [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  // Note: Default wiegand_params can be changed on the Makefile
+  wg_init(&wiegand_dev, &wiegand_param);
+
+  // Note: uncomment the following to use the defaul parameters
+  // requires to include the file "wiegand_params.h"
+  // defaults could be changed on the Makefile
+  //wg_init(&wiegand_dev, &wiegand_params);
+
+  while(1) {
+    if(wg_available(&wiegand_dev)) {
```
Comment (benpicco) [inline examples/wiegand/main.c]: Instead of polling, you could register a callback with your driver that gets called when data is available.

File: drivers/wiegand/wiegand.c [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+ * @brief Process the temporary Wiegand buffer and decodes the data.
+ *
+ */
+bool wg_do_conversion(wiegand_t *dev)
+{
+  uint64_t sysTick;
+
+  sysTick = xtimer_now_usec64();
+
+  if ((sysTick - dev->wg_last_wiegand) > 25000) // if no more signal coming through after 25ms
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c]: Instead you could do `thread_sleep()` in `wg_get_code()` and call [`xtimer_set_wakeup(timer, timeout, thread_getpid())`](https://riot-os.org/api/group__sys__xtimer.html#gab46b49ec5cf112476c83ca727bb77f67) before.

Then instead of `dev->wg_last_wiegand = xtimer_now_usec64();` you can just call `xtimer_set(timer, 25000);` to set the timeout.

File: examples/wiegand/main.c (hunk 6 above)
Comment (benpicco) [reply examples/wiegand/main.c]: Another option is to make `wg_get_code()` blocking - see my other comment.

File: drivers/include/wiegand.h (hunk 1 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: Those variables represent the wiegand buffer (*_card_temp), the last timestamp and number of bits for a wiegand frame. Those variables are modified within interrupt context.

However... After testing a little bit more I just found that **xtimer_now_usec64** is not returning a monotonic value if called within the pin change handler.

It is ok to let this "internal state" variables within the device structure? What's the best strategy to store internal state for drivers in RIOT? (I'm just trying to avoid globals)

File: drivers/include/wiegand.h (hunk 2 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: Fine. Let me do this change.

File: drivers/wiegand/wiegand.c (hunk 4 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c]: Yes. We can use only one 64 bit integer. I suppose that the original author of the algorithm wanted to make the code a little bit more readable when decoding the Mifare 34 format.

Let me do this little change.

File: drivers/include/wiegand.h (hunk 5 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: Yes. We can use only one 64 bit integer. I suppose that the original author of the algorithm wanted to make the code a little bit more readable when decoding the Mifare 34 format.

Let me do this little change.

File: examples/wiegand/main.c (hunk 6 above)
Comment (mxgxw) [reply examples/wiegand/main.c]: The indication for end-of-frame in wiegand is no-activity on any of the two data lines (D0 or D1) for more than 20ms. I was thinking originally in a callback but then I would need to start/reset a process (or a timer) that is activated *after* a level-change is detected on any of the lines.

I'm not clear for me if it's safe in RIOT to try to start/reset a timer within an interrupt context.

File: drivers/wiegand/wiegand.c (hunk 7 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c]: I didn't really wanted to make any function blocking because an RFID reader could stay without any activity during several hours. I know that the sleep strategy essentially will not use any resources but it's more just a design approach to keep the usage more readable and easy to integrate.

May be I can provide a blocking function with a timeout specified by the user.

**Comment by mxgxw** [comment]
Hi @benpicco!

Just to give you a little bit of context of this protocol.

Wiegand is a realy simple protocol that uses two data lines to signal the 0's and 1's for an RFID card ID, or a barcode, or a magnetic strip. It's a really common protocol used in access control applications. I can bet if you ever used a RFID access card (HID is one of the biggest manufacturers), the reader probably implemented this protocol. Also some numeric keypads used in access control use this protocol to transmit the keystrokes.

This driver could allow the use of such devices in RIOT. Just to give you an idea how widely used you can do an Amazon search for "Wiegand RFID reader".

Signaling is fairly simple in Wiegand. It just sends equally spaced pulses over the D0 and D1 lines to indicate the value of the bits composing the ID number of the RFID card. The end of the frame is indicated by no activity on any of those lines.

The driver was derived from an Arduino implementation. I'm cleaning up the code a little bit. Also I just finished another implementation that doesn't requires to record a timestamp on each line-level change. It works only checking if the bit-count for the wiegand frame didn't show changes between each loop. You can take a look at the new implementation [over our repo](https://github.com/hackerspacesv/RIOT/blob/drivers-hssv/wiegand/drivers/wiegand/wiegand.c#L174). 

I'll wait for your comments before porting the new algorithm over the main codebase.

Regards,
Mario.

File: drivers/include/wiegand.h (hunk 1 above)
Comment (benpicco) [reply drivers/include/wiegand.h]: Sure the driver struct is where you keep global state for the driver.
This way you can have multiple instances of the driver.

**Comment by benpicco** [comment]
@mxgxw want to PR the new implementation?
It looks fairly short, although the formatting is all over the place - maybe `uncrustify` can help.

**Comment by mxgxw** [comment]
Hello!

I'm working on some changes based on your recommendations. Unfortunately we had to close the workshop during the quarantine and has been a little bit hard to follow up. We are just re-starting to getting back to normal.

File: drivers/include/wiegand.h (hunk 1 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: Done. Included in the latest commit.

File: drivers/include/wiegand.h (hunk 2 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: Changed in the latest commit.

File: drivers/wiegand/wiegand.c (hunk 3 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c]: Done. All private functions are static now.

File: drivers/include/wiegand.h (hunk 5 above)
Comment (mxgxw) [reply drivers/include/wiegand.h]: I decided to leave it as is. Indeed the reason was to add a little bit of readability as not all cards are longer than 32 bits.

File: examples/wiegand/main.c (hunk 6 above)
Comment (mxgxw) [reply examples/wiegand/main.c]: Current implementations checks if there is a new code in the buffer each time you try to read it. Buffer is always updated when activity is detected on the line. I decided it would be better to leave the user decide how to manage the blocking instead implementing this into the driver. However I would like to know the feedback of other users to see what they think.

File: drivers/Makefile.include [hunk 8]
```diff
@@ -408,10 +408,18 @@ ifneq (,$(filter w5100,$(USEMODULE)))
...
   USEMODULE_INCLUDES += $(RIOTBASE)/drivers/ws281x/include
 endif
 
 ifneq (,$(filter xbee,$(USEMODULE)))
   USEMODULE_INCLUDES += $(RIOTBASE)/drivers/xbee/include
 endif
+
+ifneq (,$(filter xbee,$(USEMODULE)))
+  USEMODULE_INCLUDES += $(RIOTBASE)/drivers/xbee/include
+endif
```
Comment (benpicco) [inline drivers/Makefile.include:425]: This is unrelated.

File: drivers/Makefile.dep [hunk 9]
```diff
@@ -836,6 +836,12 @@ ifneq (,$(filter w5100,$(USEMODULE)))
   USEMODULE += luid
 endif
 
+ifneq (,$(filter wiegand,$(USEMODULE)))
+  USEMODULE += xtimer
```
Comment (benpicco) [inline drivers/Makefile.dep:840]: that's redundant.

File: drivers/wiegand/wiegand.c [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+    wiegand_t *dev = (wiegand_t *)arg;
+
+    // Increment bit count for Interrupt connected to D0
+    dev->wg_bit_count++;
+    // If bit count more than 31, process high bits
+    if (dev->wg_bit_count > 31) {
+        // shift value to high bits
+        dev->wg_card_temp_high |= ((0x80000000 & dev->wg_card_temp) >> 31);
+        dev->wg_card_temp_high <<= 1;
+        dev->wg_card_temp <<= 1;
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c:46]: You could also use `uint64_t wg_card_temp`, then you don't have to do this distinction.
I wonder how it affects code size though.

File: drivers/wiegand/wiegand.c [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+ * @brief Process the temporary Wiegand buffer and decodes the data.
+ *
+ */
+static bool wiegand_do_conversion(wiegand_t *dev)
+{
+    uint64_t sysTick;
+
+    sysTick = xtimer_now_usec64();
+    // Here we check if we don't have recent activity
+    // TODO: This code smells. Please fix.
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c:169]: Please do :wink:

File: drivers/wiegand/wiegand.c [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+    if (
+        sysTick - dev->wg_last_wiegand > 25000 &&
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c:171]: 25000 should better be a define

File: drivers/wiegand/wiegand.c [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+            // done 1 left shift in advance
+            dev->wg_card_temp >>= 1;
+
+            // Bit count more than 32 bits, shift high bits right to make
+            // adjustment
+            if (dev->wg_bit_count > 32) {
+                dev->wg_card_temp_high >>= 1;
+            }
+
+            if (dev->wg_bit_count == 8) { // keypress wiegand with integrity
```
Comment (benpicco) [inline drivers/wiegand/wiegand.c:198]: I think this would look cleaner if you did a 

```C
switch (dev->wg_bit_count) {
case 4:
    …
case 8:
    …
```

File: drivers/include/wiegand.h [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+ */
+typedef struct {
+    uint32_t code;                          /**< Wiegand code */
+    int16_t type;                           /**< Wiegand type */
+    bool ready;                             /**< Ready to convert */
+    int16_t last_bit_count;                 /**< Latest bit-count */
+    volatile uint32_t wg_card_temp_high;    /**< Card MSB */
+    volatile uint32_t wg_card_temp;         /**< Card LSV  */
+    volatile uint32_t wg_last_wiegand;      /**< Last card lecture */
+    volatile int16_t wg_bit_count;          /**< Number of bits read */
```
Comment (benpicco) [inline drivers/include/wiegand.h:55]: volatile won't help you with race conditions - there is no reason for this to be `volatile`.

File: drivers/wiegand/wiegand.c (hunk 12 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c:171]: Adding it as a new constant WIEGAND_WAIT under wiegand.h. Not meant to be configurable.

File: drivers/include/wiegand.h (hunk 14 above)
Comment (mxgxw) [reply drivers/include/wiegand.h:55]: Added them as volatile to prevent the compiler to optimize them at compile time. This variables are accessed by the "bit" counters at interrupt time. However looking at the code I agree that I don't think the compiler will try to do any optimizations, just trying to be extra-cautious.

File: drivers/wiegand/wiegand.c (hunk 13 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c:198]: Done. I cleaned up a little bit the whole section of the code.

File: drivers/wiegand/wiegand.c (hunk 11 above)
Comment (mxgxw) [reply drivers/wiegand/wiegand.c:169]: Done. I think is a lot better now.

File: drivers/Makefile.dep (hunk 9 above)
Comment (mxgxw) [reply drivers/Makefile.dep:840]: Removed now.

File: drivers/Makefile.include (hunk 8 above)
Comment (mxgxw) [reply drivers/Makefile.include:425]: Removed, I think it was automatically merged at some point. In any-case latest master branch removes this file.

**Comment by mxgxw** [comment]
Hi! Sorry for the radio-silence. I'm still far away from the lab. I'll try to push new changes today.

**Comment by benpicco** [comment]
Sorry for not getting back on this - still interested in the driver?
I'll try to do a review soon.
